### PR TITLE
Padding Implementation

### PR DESCRIPTION
**Pull request by owendevita**
This patch implements the random padding surrounding the BSON packet. The one included unit test ensures a proper wrapping and unwrapping of our packet. Future needs may warrant further unit testing for invalid data and edge case BSON data.

Signed-off-by: Owen De Vita <owen.n.devita@gmail.com>

### CODE REVIEW

File: src/padding.rs [hunk 1]
```diff
@@ -1,12 +1,166 @@
...
+/// Returns an error if the BSON packet is not valid (it does not start and end with the correct braces) or
+/// if there is an error during the generation of random bytes or packet encapsulation
+
+fn encapsulate(pkt: &[u8]) -> Result<Vec<u8>, std::io::Error> {
+    // if the packet does not start and end with braces, it is not valid BSON data
+    if pkt[0] != b'{' || pkt[pkt.len() - 1] != b'}' {
+        return Err(Error::from(ErrorKind::InvalidData));
+    }
+
+    let mut rng = rand_chacha::ChaCha8Rng::from_entropy(); // use the ChaCha8Rng algorithm to generate our random numbers and bytes
```
Comment (amyipdev) [inline src/padding.rs]: This was entirely unnecessary - we already have `rand` which should pull from entropy, we don't need ChaCha's randomness, especially since this is for our padding and not for cryptographic purposes.

File: src/padding.rs [hunk 2]
```diff
@@ -1,12 +1,166 @@
...
+///
+/// # Arguments
+///
+/// * `rng` - A mutable reference to a random number generator.
+///
+/// # Returns
+///
+/// A random u32 between 2 and 32 (inclusive).
+
+fn generate_random_num<R: Rng + CryptoRng>(rng: &mut R) -> u32 {
```
Comment (amyipdev) [inline src/padding.rs]: This function could be inlined, and we don't need to use Distribution - there's simpler functions we can use for pulling randomness. See the AES example in the repo for how to do that.

File: src/padding.rs [hunk 3]
```diff
@@ -1,12 +1,166 @@
...
+/// # Arguments
+///
+/// * `rng` - A mutable reference to a random number generator.
+/// * `count` - The number of bytes to generate.
+/// * `exclusion` - The character to exclude from the generated bytes.
+///
+/// # Returns
+///
+/// A vector of random bytes with the defined length, excluding the specified character.
+
```
Comment (amyipdev) [inline src/padding.rs]: Please don't have newlines between the doc comments and their associated function. Also, these doc comments are great, but they're a little overdone - in a compact view of functions we don't necessarily need this many newlines...

File: Cargo.toml [hunk 4]
```diff
@@ -6,5 +6,6 @@ edition = "2021"
 [dependencies]
 aes-gcm-siv = "0.11.1"
 rand = "0.8.5"
+rand_chacha = "0.3.1"
```
Comment (amyipdev) [inline Cargo.toml]: Remove as per lower comment

File: src/padding.rs [hunk 5]
```diff
@@ -1,12 +1,166 @@
...
+fn generate_random_bytes<R: Rng>(rng: &mut R, count: u32, exclusion: char) -> Vec<u8> {
```
Comment (amyipdev) [inline src/padding.rs]: Is this really necessary? Also, we should avoid `.expect`/`.unwrap`, instead we should return `Result`s.

File: src/padding.rs [hunk 6]
```diff
@@ -1,12 +1,166 @@
...
+/// The encapsulated packet in Vec<u8> form.
+///
+/// # Errors
+///
+/// Returns an error if the BSON packet is not valid (it does not start and end with the correct braces) or
+/// if there is an error during the generation of random bytes or packet encapsulation
+
+fn encapsulate(pkt: &[u8]) -> Result<Vec<u8>, std::io::Error> {
+    // if the packet does not start and end with braces, it is not valid BSON data
+    if pkt[0] != b'{' || pkt[pkt.len() - 1] != b'}' {
```
Comment (amyipdev) [inline src/padding.rs]: This shouldn't be here - BSON doesn't start with { and }, those are the indicators for where our padding starts/ends. Further, ensuring that the internal data is proper BSON is the responsibility of the BSON module, not the padding module.

File: src/padding.rs [hunk 7]
```diff
@@ -1,12 +1,166 @@
...
+
+    // if the bson_front_index hasn't been changed, we were unable to find an open bracket and therefore need to throw an error
+    if bson_front_index == -1 {
+        return Err(Error::from(ErrorKind::InvalidData));
+    }
+
+    // loop through the packet backwards to find the end of the BSON data
+    for (rev_index, &element) in pkt_vec.iter().rev().enumerate() {
+        if element == b'}' {
+            if rev_index < 2 || rev_index > 32 {
```
Comment (amyipdev) [inline src/padding.rs]: we should short-circuit after 32 instead of just going on forever. this also applies in the forward direction

File: src/padding.rs [hunk 8]
```diff
@@ -1,12 +1,166 @@
...
+                return Err(Error::from(ErrorKind::InvalidData));
+            } else {
+                let index = pkt_vec.len() - rev_index; // since the index is now reversed, we have to turn it back into the normal index for future use
+                bson_back_index = index as i32;
```
Comment (amyipdev) [inline src/padding.rs]: please make sure these indices don't include the braces

**Review by amyipdev — CHANGES_REQUESTED** [review]
Several small changes to be made, and several structural issues persist. Not mergeable in its current state, but shouldn't be too hard to adjust.

**Comment by amyipdev** [comment]
feel free to resolve this over however many commits, i'm gonna squash anyways
